src/common/module_builder.cc

```diff
@@ -65,6 +66,36 @@ ModuleBuilder::ModuleBuilder()
   m_context->appendDialectRegistry(registry);
 }
 
+static bool isScalarType(mlir::Type type) {
+  if (mlir::isa<mlir::FloatType>(type) || mlir::isa<mlir::IntegerType>(type)) {
+    return true;
+  }
+  if (auto tensorType = mlir::dyn_cast<mlir::RankedTensorType>(type)) {
+    return tensorType.getRank() == 0;
+  }
+  return false;
+}
+
+void ModuleBuilder::collectOutputTypes(mlir::ModuleOp &&module) {
+  m_is_output_scalar.clear();
+  for (auto &op : module.getOps()) {
+    if (auto funcOp = mlir::cast<mlir::func::FuncOp>(op)) {
+      // We care only for return ops of public functions, as that are the ones
+      // that will produce results in the flatbuffer.
+      if (funcOp.isPublic()) {
```

Comment:
What does it mean for op to be public?

Reply:
Function will be public if it was compiled to be directly used by the frontend, and hence will have output buffers, which means having output buffers. If the function is related to some jax optimisation, compilation specific (as is the `jax.lax` remainder case), it will be private.